### PR TITLE
bpo-40346: Random subclass randbytes uses getrandbits

### PR DESCRIPTION
**Pull request by vstinner**
Subclass of the random.Random class now get a randbytes() method
implementation which uses the getrandbits() method.

https://bugs.python.org/issue40346

### CODE REVIEW

**Comment by vstinner** [comment]
cc @pitrou @serhiy-storchaka @rhettinger

**Comment by vstinner** [comment]
I didn't change the documentation. Or do you think that it must be said explicitly that Random subclass implement randbytes() with getrandbits()? My minor documentation issue is that SystemRandom subclasses don't get a randbytes() implementation which uses getrandbits(): SystemRandom subclasses get SystemRandom.randbytes() which calls os.urandom().

**Comment by vstinner** [comment]
> I didn't change the documentation. (...) My minor documentation issue is that SystemRandom subclasses don't get a randbytes() implementation which uses getrandbits() (...)

Oh, I found a project in the wild which creates a SystemRandom subclass:
https://pypi.org/project/random-xe/

So I modified my PR to change also SystemRandom subclasses, and I completed the documentation.

File: Lib/random.py [hunk 1]
```diff
@@ -120,6 +120,11 @@ def __init_subclass__(cls, /, **kwargs):
                 cls._randbelow = cls._randbelow_without_getrandbits
                 break
 
+        if cls.randbytes == _random.Random.randbytes:
```
Comment (serhiy-storchaka) [inline Lib/random.py]: I would add the test for `getrandbits`. No need to override `randbytes` if `getrandbits` is not overriden.

Also, this does not work good for
```python
class R(Random):
    def getrandbits(self, n): ...
    randbytes = Random.randbytes
```

File: Lib/random.py [hunk 2]
```diff
@@ -189,6 +194,13 @@ def setstate(self, state):
                              "Random.setstate() of version %s" %
                              (version, self.VERSION))
 
+## -------------------- bytes -----------------------------
+
+    # Implementation used by Random and SystemRandom subclasses
+    def _randbytes_getrandbits(self, n):
```
Comment (serhiy-storchaka) [inline Lib/random.py]: `__name__` and `__qualname__` of this method does not look nice.

You can define method with name `randbytes` and then rename it.
```python
def randbytes(self, n): ...
_randbytes_getrandbits = randbytes
del randbytes
```
Or better create an internal `randbytes` function in `__init_subclass__` and set its `__qualname__` (which depends on the class name) manually.

File: Lib/random.py (hunk 1 above)
Comment (vstinner) [reply Lib/random.py]: Fixed

File: Lib/random.py (hunk 2 above)
Comment (vstinner) [reply Lib/random.py]: Fixed. I also added a test.

**Comment by vstinner** [comment]
I updated the PR to address @serhiy-storchaka's comments and I added more tests and two other kinds of subclasses.

File: Lib/random.py [hunk 3]
```diff
@@ -732,6 +749,15 @@ class SystemRandom(Random):
      Not available on all systems (see os.urandom() for details).
     """
 
+    def __init_subclass__(cls, /, **kwargs):
+        super.__init_subclass__(**kwargs)
```
Comment (serhiy-storchaka) [inline Lib/random.py]:
```suggestion
        super().__init_subclass__(**kwargs)
```

File: Lib/random.py [hunk 4]
```diff
@@ -732,6 +749,15 @@ class SystemRandom(Random):
...
+
+        if (cls.randbytes == SystemRandom.randbytes
+           and cls.getrandbits != _random.Random.getrandbits):
```
Comment (serhiy-storchaka) [inline Lib/random.py]:
```suggestion
           and cls.getrandbits != _random.SystemRandom.getrandbits):
```

**Comment by serhiy-storchaka** [comment]
I am not sure that pickling the `randbytes` method of `Random` subclass works correctly. If it does not work it may be better to use the second option for the `randbytes` implementation.

**Comment by vstinner** [comment]
> I am not sure that pickling the randbytes method of Random subclass works correctly. If it does not work it may be better to use the second option for the randbytes implementation.

pickle can be used to serialize/deserialize instances. `__init_subclass__()` is called when a new class is created.

Example:

```
import random
import pickle

class MyRandom(random.Random):
    def __init__(self):
        super().__init__()
        self.getrandbits_calls = []
        print("init", self)

    def getrandbits(self, n):
        self.getrandbits_calls.append(n)
        return 0

rng = MyRandom()
print(rng.randbytes != random.Random.randbytes)
print()
rng2 = pickle.loads(pickle.dumps(rng))
print(rng2.randbytes != random.Random.randbytes)
```

Output:

```
init <__main__.MyRandom object at 0xefaff0>
True

init <__main__.MyRandom object at 0xf2f320>
True
```

Two MyRandom instances are created, but MyRandom.randbytes is only set once by Random.__init_subclass(MyRandom): when the class is created.

**Comment by serhiy-storchaka** [comment]
I meant pickling the method.
```python
pickle.dumps(MyRandom.randbytes)
```

**Comment by vstinner** [comment]
> I meant pickling the method.
> pickle.dumps(MyRandom.randbytes)

Alright, that's broken because you asked me to ensure that `Random._randbytes_getrandbits` is called "randbytes". I reverted this change, now it works again. Example:

```
import random
import pickle

class MyRandom(random.Random):
    def __init__(self):
        super().__init__()

    def getrandbits(self, n):
        return 0

rng = MyRandom()
print(MyRandom.randbytes is random.Random._randbytes_getrandbits)
randbytes2 = pickle.loads(pickle.dumps(MyRandom.randbytes))
print(randbytes2 is random.Random._randbytes_getrandbits)
```

Output:

```
True
True
```

**Comment by serhiy-storchaka** [comment]
I think we can have this cake and eat it. Define `randbytes()` in `__init_subclass__()` and set its `__module__` and `__qualname__`.

**Comment by vstinner** [comment]
@serhiy-storchaka:
> If it does not work it may be better to use the second option for the randbytes implementation.

Which second option? Remove the C implementation and always implement randbytes() using getrandbits() in Lib/random.py?

Well, I proposed a solution to keep the optimization (C implementation) and don't require subclasses to have to override a 6th method (randbytes).

I'm not super excited by adding more special cases `__init_subclass__()` (this PR). I prefer PR #19631 (random.BaseRandom) approach which uses more "regular Python code".

I let @rhettinger and you decide which approach is the best.

When I read https://bugs.python.org/issue40346 it's not even clear if we support subclassing or not, nor if performance matters or not. I'm now very confused :-( Now at least, we have multiple choices, and I let you decide which one is the best ;-)

**Comment by serhiy-storchaka** [comment]
The second option which I mentioned in my comment https://github.com/python/cpython/pull/19700#discussion_r414490521 and repeated in https://github.com/python/cpython/pull/19700#issuecomment-619123719.

**Comment by vstinner** [comment]
> create an internal randbytes function in __init_subclass__ and set its __qualname__ (which depends on the class name) manually.

Well, it seems like there are different ways to implement this change. But I'm not sure that this approach is the best. For now, I will leave the PR as it is.

I'm now waiting for a feedback on https://bugs.python.org/issue40346#msg367202

**Comment by vstinner** [comment]
I abandon my BaseRandom PR #19631 and this PR since there is no clear consensus on these changes: see https://bugs.python.org/issue40346

I wrote PR #19797 instead: "Remove C implementation of Random.randbytes()".

**Pull request closed without merging**